### PR TITLE
v1.0.0-beta

### PR DESCRIPTION
**Pull request by max-degterev**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  new I18nModular(),
+],
+```
+
+Create an options RC file in your project's root named `.i18n-modular-rc.js` with this example content:
+```javascript
+module.exports = {
+  // Opional, your keys will be relative to this folder, usually `./app` or `./client`
+  keysRoot: './',
+  // Opional, how you will be naming your translation modules, similar to name.module.css
```
Comment (peterhass) [inline README.md]: Op**t**ional

File: lib/loaders/seed.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+  const { store } = options;
+
+  const language = getLanguage(this.resourcePath.replace(this.context, ''));
+
+  const id = getId(language);
+  const data = JSON.parse(source);
+
+  // Stops webpack from watching changes to the dictionary.
+  // We need this since we read from and output to the same file.
```
Comment (peterhass) [inline lib/loaders/seed.js]: ... since we read and write to the same file

File: README.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
+i18n-modular
+============
+
+A webpack plugin that brings support for a modular approach to working with dictionaries. Especially useful for tools such as [Phraseapp](https://phrase.com). Same as CSS Modules, but for i18n.
+
+## Usage:
```
Comment (peterhass) [inline README.md:6]: Usage section is missing on how to import .translation files

File: lib/plugin.js [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+  getDictionary(id) {
+    const { language, data } = this.seeds.get(id);
+    return getSortedObject({ ...getCleanSeed(data), ...this.getModules(language) });
+  }
+
+  apply(compiler) {
+    const { moduleEnding, dictionaryPattern } = this.options;
+    const { path: compilerOutputPath } = compiler.options.output;
+    const outputPattern = getOutputPattern(dictionaryPattern || compilerOutputPath);
+    const dictionaryRegex = getDictionaryRegex(outputPattern.replace(compiler.context, ''));
```
Comment (peterhass) [inline lib/plugin.js]: extract `outputPattern.replace(compiler.context, '')` to a variable to improve readability?

File: lib/plugin.js [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    });
+
+    compiler.hooks.emit.tapAsync(PLUGIN_NAME, (compilation, callback) => {
+      debug('building dictionaries');
+
+      Array.from(this.seeds.keys()).forEach((id) => {
+        const { language } = this.seeds.get(id);
+        const filePath = getOutputPath(outputPattern, language);
+
+        // Webpack is unable to save asseds to absolute paths, a relative to the output folder
```
Comment (peterhass) [inline lib/plugin.js]: `assets`

**Review by peterhass — APPROVED** [review]
cool

**Review by peterhass — APPROVED** [review]

File: lib/loaders/module.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  keys.forEach((key) => {
+    if (!REGEX_LOCALE.test(key)) throw new Error(`Language key ${key} in ${filePath} is invalid`);
+  });
+};
+
+const loader = function(source) {
+  const options = getOptions(this);
+  validate(schema, options, { name: 'I18nModular module loader' });
+
+  const { store, getId } = options || { getId: getDefaultId };
```
Comment (atomiomi) [inline lib/loaders/module.js:51]: getId is required option? can be null of not passed in options

File: lib/sync.js [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+  getCleanGenerated,
+
+  getSortedObject,
+  getDeepSortedObject,
+} = require('./utils');
+
+
+const getFiles = (dirPath, regex) => (
+  fs.readdirSync(dirPath).reduce((acc, fileName) => {
+    const filePath = path.resolve(`${dirPath}/${fileName}`);
```
Comment (atomiomi) [inline lib/sync.js:25]: `resolve(dirPath, fileName)`

File: lib/loaders/module.js [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+  const id = getId(this.resourcePath);
+  const data = JSON.parse(source);
```
Comment (atomiomi) [inline lib/loaders/module.js:54]: does webpack throw an error if source is invalid json?

Comment (max-degterev) [reply lib/loaders/module.js:54]: JSON.parse would throw and webpack should catch

File: lib/sync.js (hunk 7 above)
Comment (max-degterev) [reply lib/sync.js:25]: same thing, it's normalized

File: lib/loaders/module.js (hunk 8 above)
Comment (peterhass) [reply lib/loaders/module.js:54]: I tried it. Webpack formats the error message pretty good

**Pull request merged**